### PR TITLE
v3.0.0 - React hooks API

### PR DESCRIPTION
**Pull request by fabien0102**
## Summary

Re-write the entire library to return a React hook API instead of Render props components.

Since the last update of `react-runtime-config` was a long time ago, this PR also fresh up the stack!

The main API is totally new, to have more runtime type safety, more information about this in the update `README.md`

### CODE REVIEW

File: .husky/.gitignore [hunk 1]
```diff
@@ -0,0 +1 @@
+_
```
Comment (mpotomin) [inline .husky/.gitignore:1]: 👀

Comment (fabien0102) [reply .husky/.gitignore:1]: Husky v6 pattern ¯\_(ツ)_/¯

File: README.md [hunk 2]
```diff
@@ -78,54 +63,64 @@ This configuration is then easily read by the simple React component that this l
...
+      min: 1,
+      max: 65535,
+      default: 8000, // config with `default` don't have to be set on `window.MY_APP_CONFIG`
+    },
+    monitoringLink: {
+      type: "custom",
+      description: "Link of the monitoring",
+      parser: value => {
+        if (typeof value === "object" && typeof value.url === "string" && typeof value.displayName === "string") {
+          // The type will be infered from the return type
```
Comment (mpotomin) [inline README.md:95]:
```suggestion
          // The type will be inferred from the return type
```

File: README.md [hunk 3]
```diff
@@ -78,54 +63,64 @@ This configuration is then easily read by the simple React component that this l
...
+          return { url: value.url as string, displayName: value.displayName as string };
+        }
+        // This error will be shown if the `window.MY_APP_CONFIG.monitoringLink` can't be parsed or if we `setConfig` an unvalid value
```
Comment (mpotomin) [inline README.md:98]:
```suggestion
        // This error will be shown if the `window.MY_APP_CONFIG.monitoringLink` can't be parsed or if we `setConfig` an invalid value
```

File: README.md [hunk 4]
```diff
@@ -134,84 +129,123 @@ The priority of config values is as follows:
...
 You have also access to `field.windowValue` and `field.storageValue` if you want implement more advanced UX on this page.
 
+## Multiconfiguration admin page
+
+As soon as you have more than one configuration in your project, creating an administration page than merge everything is a nice to have. Of course, you will want a kind of `ConfigSection` component that take the result of any `useAdminConfig()` (so `field`, `reset` and `namespace` as props).
+
+Spoiler alert, having this kind of component type safe can be tricky, indeed you can try use `ReturnType<typeof useFirstAdminConfig> | ReturnType<typeof useSecondAdminConfig>` as props but typescript will fight you (`Array.map` will tell you that the signature are not compatible).
+
+Anyway, long story short, this library provide you an easy way to with this: `GenericAdminFields` type. This type is compatible with every configuration and will provide you a nice framework to create an amazing UX.
+
```
Comment (mpotomin) [inline README.md:209]: Since it is a public lib could be nice to have a native speaker proof-read the doc update :)

File: src/createUseConfig.ts [hunk 5]
```diff
@@ -0,0 +1,20 @@
+import { Config, InjectedProps, NamespacedUseConfigReturnType } from "./types";
+import { useCallback } from "react";
+import { useWatchLocalstorageEvents, capitalize } from "./utils";
+
+export function createUseConfig<TSchema extends Record<string, Config>, Namespace extends string>(
+  props: InjectedProps<TSchema, Namespace>,
+) {
+  return () => {
+    const localstorageDep = useWatchLocalstorageEvents(props.storage, props.localOverride);
```
Comment (mpotomin) [inline src/createUseConfig.ts:9]: For the sake of consistency with the `window` propery name, let's name it camelcase here as well 
```suggestion
    const localStorageDependency = useWatchLocalstorageEvents(props.storage, props.localOverride);
```

File: src/utils.ts [hunk 6]
```diff
@@ -1,47 +1,30 @@
...
-      e: infer E,
-      f: infer F,
-      g: infer G,
-      h: infer H,
-      i: infer I,
-      j: infer J,
-    ) => any
-  ? [A, B, C, D, E, F, G, H, I, J]
-  : never[];
+export const useWatchLocalstorageEvents = (storage: Storage, localOverride: boolean) => {
```
Comment (mpotomin) [inline src/utils.ts:9]: Same regarding the naming here
```suggestion
export const useWatchLocalStorageEvents = (storage: Storage, localOverride: boolean) => {
```

File: src/types.ts [hunk 7]
```diff
@@ -0,0 +1,221 @@
+export interface ConfigOptions<TSchema extends Record<string, Config>, TNamespace extends string> {
+  /**
+   * Namespace of the configuration
+   *
+   * This namespace is used to consume the configuration from `window` and `localstorage`
```
Comment (mpotomin) [inline src/types.ts:5]:
```suggestion
   * This namespace is used to consume the configuration from `window` and `localStorage`
```

File: src/types.ts [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+
+  /**
+   * Storage adapter
+   *
+   * @default window.localStorage
+   */
+  storage?: Storage;
+
+  /**
+   * Permit to overidde any config values in storage
```
Comment (mpotomin) [inline src/types.ts:22]:
```suggestion
   * Permit to override any config values in storage
```

File: src/types.ts [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+  /**
+   * Value stored in `window.{path}`
+   */
+  windowValue: ResolvedConfigValue<TSchema[TKey]> | null;
+  /**
+   * Value stored in `storage.getItem({path})`
+   */
+  storageValue: ResolvedConfigValue<TSchema[TKey]> | null;
+  /**
+   * True if a value is store on the localstorage
```
Comment (mpotomin) [inline src/types.ts:150]:
```suggestion
   * True if a value is store on the localStorage
```

File: src/types.ts [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+  default?: string | (() => string);
+  description?: string;
+}
+
+export interface StringEnumConfig extends StringConfig {
+  /**
+   * List of allowed values
+   */
+  enum: string[];
+}
```
Comment (mpotomin) [inline src/types.ts:63]: If `StringEnumConfig` extends `StringConfig` with non-optional `enum`, why `StringConfig` already has `enum?: string[];`.

File: src/types.ts [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+  storageValue: T | null;
+  isFromStorage: boolean;
+  value: T;
+  set: (value: U) => void;
+};
+
+/**
+ * `useAdminConfig.fields` in a generic version.
+ *
+ * This should be used if you are implement a generic component
```
Comment (mpotomin) [inline src/types.ts:188]:
```suggestion
 * This should be used if you are implementing a generic component
```

File: src/types.ts [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+ * that consume any `fields` as prop.
+ *
+ * Note: "custom" type and "string" with enum are defined as `any` to be
+ * compatible with any schemas. You will need to validate them in your
+ * implementation to retrieve type safety.
```
Comment (mpotomin) [inline src/types.ts:193]:
```suggestion
 * implementation to retrieve type.
```

File: src/types.ts [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+  setConfig: <K extends keyof TSchema>(path: K, value: ResolvedConfigValue<TSchema[K]>) => void;
+};
+
+type Namespaced<T, TNamespace extends string> = {
+  [P in keyof T as P extends `getConfig`
+    ? `get${Capitalize<TNamespace>}Config`
+    : P extends "getAllConfig"
+    ? `getAll${Capitalize<TNamespace>}Config`
+    : `set${Capitalize<TNamespace>}Config`]: T[P];
+};
```
Comment (mpotomin) [inline src/types.ts:216]: I have hard time understanding what is happening here. And tripple ternary does not make it easier :)
An asdoc description could be helpful

File: src/types.ts (hunk 10 above)
Comment (fabien0102) [reply src/types.ts:63]: So you can do `field.type === "string" && field.enum`, otherwise this is painful to discriminate

need `as any` to access `enum:
image

`enum` is not listed in the `field` properties
image

File: src/utils.ts [hunk 14]
```diff
@@ -1,47 +1,30 @@
...
+  const [key, setKey] = React.useState(0);
+
+  React.useEffect(() => {
+    const onStorageUpdate = (_: StorageEvent) => {
+      if (storage && localOverride) {
+        setKey(i => (i + 1) % 10);
```
Comment (mpotomin) [inline src/utils.ts:15]: Would be nice to have a description why we need modulo of 10 here

File: src/utils.ts [hunk 15]
```diff
@@ -1,47 +1,30 @@
...
+    return () => window.removeEventListener("storage", onStorageUpdate);
+  }, [storage, localOverride, setKey]);
+
+  return key;
+};
+
+export function capitalize<T extends string>(str: T) {
+  if (!str || str.length < 1) return "" as Capitalize<T>;
+  return (str[0].toUpperCase() + str.slice(1)) as Capitalize<T>;
+}
```
Comment (mpotomin) [inline src/utils.ts:30]: This util seem to be not related to the previous one in same file. Would it make sense to move it to a separate file to avoid confusion? Both files could live within `utils` folder

File: src/createUseAdminConfig.ts [hunk 16]
```diff
@@ -0,0 +1,50 @@
+import { InjectedProps, Config, ResolvedConfigValue, AdminFields } from "./types";
+import { useCallback, useMemo } from "react";
+import { useWatchLocalstorageEvents } from "./utils";
+
+export function createUseAdminConfig<T extends Record<string, Config>, TNamespace extends string>(
```
Comment (mpotomin) [inline src/createUseAdminConfig.ts:5]: To clarify the `T` type, could we rename `T` to `TConfig` or `TNamespacedConfig` here? It will also align it with `TNamespace`

File: src/createUseAdminConfig.ts [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+  props: InjectedProps<T, TNamespace>,
+) {
+  return () => {
+    const localstorageDep = useWatchLocalstorageEvents(props.storage, props.localOverride);
```
Comment (mpotomin) [inline src/createUseAdminConfig.ts:9]:
```suggestion
    const localStorageDep = useWatchLocalStorageEvents(props.storage, props.localOverride);
```

File: src/createUseAdminConfig.ts (hunk 16 above)
Comment (mpotomin) [reply src/createUseAdminConfig.ts:5]: or is it rather a `TSchema` ?

File: src/parsers.ts [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+    return parseFloat(value);
+  }
+  throw new Error("not a number");
+}
+
+function parseBoolean(value: unknown): boolean {
+  if (typeof value === "boolean") {
+    return value;
+  }
+  if (typeof value === "string" && ["true", "false"].includes(value.toLowerCase())) {
```
Comment (mpotomin) [inline src/parsers.ts:35]: Good logic 👍  Is there a test for it?

File: src/types.ts (hunk 10 above)
Comment (mpotomin) [reply src/types.ts:63]: Could adding a separate field type `stringWithEnum` make it a bit clearer?

File: src/parsers.ts (hunk 18 above)
Comment (fabien0102) [reply src/parsers.ts:35]: https://github.com/contiamo/react-runtime-config/blob/a333f6599d254b18df2d82343a8f3a273fdd7199/src/index.test.ts#L243-L255

File: src/index.ts [hunk 19]
```diff
@@ -0,0 +1,179 @@
...
+  isStringEnumConfig,
+  isNumberConfig,
+  isCustomConfig,
+  isBooleanConfig,
+  isStringConfig,
+  // useConfigAdmin.fields
+  AdminField,
+  AdminFields,
+  GenericAdminFields,
+};
```
Comment (mpotomin) [inline src/index.ts:46]: Can commented `Options`, `Configs`, `Typeguards` etc be removed from here?

Comment (fabien0102) [reply src/index.ts:46]: This was to organized a bit the exports (and avoid to forgot some ^^)

Comment (fabien0102) [reply src/index.ts:46]: (It's a title, not a commented type)

File: src/types.ts [hunk 20]
```diff
@@ -0,0 +1,221 @@
...
+export const isBooleanConfig = (config: Config): config is BooleanConfig => config.type === "boolean";
+export const isCustomConfig = (config: Config): config is CustomConfig => config.type === "custom";
+
+export interface InjectedProps<
+  TSchema extends Record<string, Config>,
+  TNamespace extends string,
+  TConfig = ResolvedSchema<TSchema>
+> {
+  namespace: string;
+  useConfigNamespace: TNamespace;
```
Comment (mpotomin) [inline src/types.ts:120]: `use` prefix make it easy to confuse with a hook. Would naming it 
```suggestion
configNamespace: TNamespace;
```
convey enough needed context?

File: src/index.test.ts [hunk 21]
```diff
@@ -0,0 +1,448 @@
...
+      },
+      isAwesome: true,
+    });
+  });
+  afterEach(() => {
+    act(() => storage.clear());
+    delete (window as any)[namespace];
+  });
+
+  it("should throw if a window value don't fit the schema", () => {
```
Comment (mpotomin) [inline src/index.test.ts:118]:
```suggestion
  it("should throw if a window value doesn't fit the schema", () => {
```

File: src/index.test.ts [hunk 22]
```diff
@@ -0,0 +1,448 @@
...
+    });
+
+    it("should return the default value (function)", () => {
+      let port = 8000;
+      const getPort = () => port;
+      const { getConfig } = createConfig({
+        namespace,
+        storage,
+        schema: {
+          port: {
```
Comment (mpotomin) [inline src/index.test.ts:138]: Are we missing `parser` part here? 

Screenshot 2021-04-15 at 16 12 46

File: src/index.test.ts [hunk 23]
```diff
@@ -0,0 +1,448 @@
...
+    });
+
+    it("should return a custom parsed value", () => {
+      const { getConfig } = createConfigWithDefaults();
+      const monitoringLink = getConfig("monitoringLink");
+      expect(monitoringLink.url).toBe("http://localhost:5000");
+      expect(monitoringLink.displayName).toBe("Monitoring");
+    });
+
+    it("should return the localstorage value (storage set before)", () => {
```
Comment (mpotomin) [inline src/index.test.ts:161]:
```suggestion
    it("should return the localStorage value (storage set before)", () => {
```

File: src/index.test.ts [hunk 24]
```diff
@@ -0,0 +1,448 @@
...
+      storage.setItem(`${namespace}.color`, "pink");
+      const { getConfig } = createConfigWithDefaults();
+      expect(getConfig("color")).toBe("pink");
+    });
+
+    it("should return the localstorage value (storage set after)", () => {
```
Comment (mpotomin) [inline src/index.test.ts:167]:
```suggestion
    it("should return the localStorage value (storage set after)", () => {
```

File: src/index.test.ts [hunk 25]
```diff
@@ -0,0 +1,448 @@
...
+      const { getConfig } = createConfigWithDefaults();
+      storage.setItem(`${namespace}.color`, "pink");
+      expect(getConfig("color")).toBe("pink");
+    });
+
+    it("should return the localstorage value (with setConfig)", () => {
```
Comment (mpotomin) [inline src/index.test.ts:173]:
```suggestion
    it("should return the localStorage value (with setConfig)", () => {
```

File: src/index.test.ts [hunk 26]
```diff
@@ -0,0 +1,448 @@
...
+    });
+
+    it("should set a value (boolean=false)", () => {
+      const { getConfig, setConfig } = createConfigWithDefaults();
+      expect(getConfig("isAwesome")).toBe(true);
+      setConfig("isAwesome", false);
+      expect(getConfig("isAwesome")).toBe(false);
+    });
+
+    it("should remove the localstorage value if same as the window one", () => {
```
Comment (mpotomin) [inline src/index.test.ts:257]:
```suggestion
    it("should remove the localStorage value if same as the window one", () => {
```

File: src/index.test.ts [hunk 27]
```diff
@@ -0,0 +1,448 @@
...
+      // Add a custom value
+      setConfig("isAwesome", false);
+      expect(storage.getItem("test.isAwesome")).toBe("false");
+
+      // Set back the default value
+      setConfig("isAwesome", true);
+      expect(storage.getItem("test.isAwesome")).toBe(null);
+    });
+
+    it("should remove the localstorage value if same as the default one", () => {
```
Comment (mpotomin) [inline src/index.test.ts:268]:
```suggestion
    it("should remove the localStorage value if same as the default one", () => {
```

File: src/index.test.ts [hunk 28]
```diff
@@ -0,0 +1,448 @@
...
+
+    it("should throw if the value is not respecting a custom parser", () => {
+      const { setConfig } = createConfigWithDefaults();
+      expect(() => setConfig("monitoringLink", "red" as any)).toThrowErrorMatchingInlineSnapshot(
+        `"Monitoring link invalid!"`,
+      );
+    });
+  });
+
+  describe("useConfig", () => {
```
Comment (mpotomin) [inline src/index.test.ts:315]: Can we spit big test file into smaller ones for easier navigation?

File: src/index.test.ts [hunk 29]
```diff
@@ -0,0 +1,448 @@
...
+      act(() => setBoomConfig("oh", "popopo"));
+      expect(getAllBoomConfig()).toMatchInlineSnapshot(`
+        Object {
+          "oh": "popopo",
+        }
+      `);
+    });
+  });
+
+  describe("useAdminConfig", () => {
```
Comment (mpotomin) [inline src/index.test.ts:376]: Same here, a separate file would be nice

File: src/index.test.ts (hunk 22 above)
Comment (fabien0102) [reply src/index.test.ts:138]: image

Strange… Are you on the correct TS version (I also try with `tsc` everything is compiling)

**Review by mpotomin — APPROVED** [review]
Great job! 💪
